src/Raven.Server/ServerWide/Commands/StartBucketMigrationCommand.cs

```diff
@@ -0,0 +1,181 @@
+using Raven.Client.ServerWide;
+using Raven.Server.Utils;
+using Sparrow.Json.Parsing;
+
+namespace Raven.Server.ServerWide.Commands
+{
+    public class StartBucketMigrationCommand : UpdateDatabaseCommand
+    {
+        public int SourceShard;
+        public int DestinationShard;
+        public int Bucket;
+
+        public StartBucketMigrationCommand()
+        {
+        }
+
+        public StartBucketMigrationCommand(string database, string raftId) : base(database, raftId)
+        {
+            
+        }
+        public override void UpdateDatabaseRecord(DatabaseRecord record, long etag)
+        {
+            if (record.BucketMigrations.Count > 0)
+            {
+                foreach (var migration in record.BucketMigrations)
+                {
+                    if (migration.Value.Status < MigrationStatus.OwnershipTransferred)
+                        throw new InvalidOperationException(
+                            $"Only one bucket can be transferred at a time, currently bucket {migration.Key} is {migration.Value.Status}");
+
+                    if (migration.Key == Bucket)
+                        throw new InvalidOperationException($"Can't migrate bucket {Bucket}, since it is still migrating.");
+                }
+            }
+
+            record.BucketMigrations.Add(Bucket, new BucketMigration
+            {
+                Bucket = Bucket,
+                DestinationShard = DestinationShard,
+                SourceShard = SourceShard,
+                MigrationIndex = etag,
+                Status = MigrationStatus.Moving
+            });
+        }
+
+        public override void FillJson(DynamicJsonValue json)
+        {
+            json[nameof(SourceShard)] = SourceShard;
+            json[nameof(DestinationShard)] = DestinationShard;
+            json[nameof(Bucket)] = Bucket;
+        }
+    }
+
+    public class SourceMigrationSendCompletedCommand : UpdateDatabaseCommand
+    {
+        public int Bucket;
+        public long MigrationIndex;
+
+        public string LastSentChangeVector;
+
+        public SourceMigrationSendCompletedCommand()
+        {
+
+        }
+
+        public SourceMigrationSendCompletedCommand(string database, string raftId) : base(database, raftId)
+        {
+            
+        }
+        public override void UpdateDatabaseRecord(DatabaseRecord record, long etag)
+        {
+            if (record.BucketMigrations.TryGetValue(Bucket, out var migration) == false)
+                throw new InvalidOperationException($"Bucket '{Bucket}' not found in the migration buckets");
+
+            if (migration.MigrationIndex != MigrationIndex)
+                throw new InvalidOperationException($"Wrong migration index. Expected: '{MigrationIndex}', Actual: '{migration.MigrationIndex}'");
+
+            if (migration.Status != MigrationStatus.Moving)
+                throw new InvalidOperationException($"Expected status is '{MigrationStatus.Moving}', Actual '{migration.Status}'");
+
```

Comment:
How do you recover from this?

Reply:
manually if needed. Note that those exceptions wouldn't jam the state machine. They will be caught and rethrow them as `RachisApplyException`